src/Refitter.Core/SchemaCleaner.cs

```diff
@@ -17,18 +19,39 @@ public SchemaCleaner(OpenApiDocument document, string[] keepSchemaPatterns)
 
     public void RemoveUnreferencedSchema()
     {
-        var usage = FindUsedSchema(document);
-
+        var (usedJsonSchema, usage) = FindUsedJsonSchema(document);
         var unused = document.Components.Schemas.Where(s => !usage.Contains(s.Key))
             .ToArray();
 
         foreach (var unusedSchema in unused)
         {
             document.Components.Schemas.Remove(unusedSchema);
         }
+
+        if (!IncludeInheritanceHierarchy)
+        {
+            foreach (var schema in usedJsonSchema)
+            {
+                // Fix any "abstract/sum" types so that the unused-types get removed
+                if (schema.DiscriminatorObject != null)
+                {
+                    var mappings = schema.DiscriminatorObject.Mapping;
+                    var keepMappings = mappings.Where(x =>
+                            usage.Contains(x.Value.ActualSchema.Id ?? x.Value.Id ?? "_________SOME NON EXISTING VALUE")
```

Comment:
I'm not sure I understand why you check for `_________SOME NON EXISTING VALUE` 😄

Reply:
I missed the ***WIP*** in the description. I will wait with going through until you say its ready for review @kirides

Reply:
Just some nullability shenanigans, could just as well extract the IDs, check for existence and only then so the contains.